### PR TITLE
Feature/dynamic field mapping

### PR DESCRIPTION
**Pull request by michaloo**

### CODE REVIEW

File: manifest.json [hunk 1]
```diff
@@ -66,6 +66,36 @@
...
+        "properties" : {
+          "hull" : {
+            "type" : "string",
+            "format" : "trait",
+            "title" : "Hull Attribute"
+          },
+          "name" : {
+            "type" : "string",
+            "format": "select",
+            "title" : "Mailchimp Attribute",
```
Comment (sbellity) [inline manifest.json]: Let's call that Merge field since that's the terminology that they use...

File: server/lib/sync-agent/index.js [hunk 2]
```diff
@@ -49,16 +49,9 @@ export default class SyncAgent {
...
-      // type check, empty fields, fields that doesn't exist?
-      // change the check if the users was already synced (update the traits)
-      // sync from hull -> mailchimp
       return {
         email_type: "html",
-        merge_fields: {
-          FNAME: user.first_name || "",
-          LNAME: user.last_name || ""
-        },
+        merge_fields: this.userMappingAgent.getMailchimpAttrbutes(user),
```
Comment (sbellity) [inline server/lib/sync-agent/index.js]: getMailchimpAttrbutes -> getMailchimpAttributes

typo ?

File: server/lib/sync-agent/user-mapping-agent.js [hunk 3]
```diff
@@ -108,4 +108,38 @@ export default class UserMappingAgent {
...
+   */
+  getMailchimpAttrbutes(hullUser) {
+    const mailchimpAttrbutes = _.reduce(this.computeMailchimpAttributes(), (fields, prop) => {
+      if (_.get(hullUser, prop.hull)) {
+        _.set(fields, prop.name, _.get(hullUser, prop.hull));
+      }
+      return fields;
+    }, {});
+
+    return mailchimpAttrbutes;
```
Comment (sbellity) [inline server/lib/sync-agent/user-mapping-agent.js]: `mailchimpAttrbutes` -> `mailchimpAttributes` 
typo ?

File: server/lib/sync-agent/index.js (hunk 2 above)
Comment (sbellity) [inline server/lib/sync-agent/index.js]: FNAME and LNAME are actually mandatory on Mailchimp right ?
I think the user update operation will fail if they are not present

Comment (michaloo) [inline server/lib/sync-agent/index.js]: Thanks for pointing out the typo.

Regarding `FNAME` and `LNAME`, it looks like it depends on the list configuration:
screen shot 2017-01-06 at 13 35 38
Having the configuration as above I can sync user with email only.

@unity said we don't need any default mapping here.

File: server/actions/schema-user-fields.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
+
+export default function schemaUserFields(req, res) {
+  req.shipApp.mailchimpAgent.getMergeFields()
+    .then(resBody => {
+      res.json({
```
Comment (michaloo) [inline server/actions/schema-user-fields.js]: Not yet, I will push it in a minute.

File: manifest.json (hunk 1 above)
Comment (michaloo) [inline manifest.json]: Let's have the discussion on trello.

**Comment by michaloo** [comment]
[Hull -> Mailchimp Merge fields mapping](https://trello.com/c/mXUaSzC9/122-hull-mailchimp-merge-fields-mapping)

**Comment by michaloo** [comment]
@sbellity - changes applied :)

File: server/lib/sync-agent/user-mapping-agent.js [hunk 5]
```diff
@@ -108,4 +108,42 @@ export default class UserMappingAgent {
...
+  /**
+   * Get ready object of attributes and its values converted from Hull
+   * user to Mailchimp list member
+   *
+   * @returns {Object}
+   */
+  getMergeFields(hullUser) {
+    const mailchimpAttributes = _.reduce(this.computeMergeFields(), (fields, prop) => {
+      if (_.get(hullUser, prop.hull)) {
+        if (!_.get(prop, "overwrite")) {
```
Comment (michaloo) [inline server/lib/sync-agent/user-mapping-agent.js]: I've also added support for `overwrite` setting to let the user have greater control over dataflow.

**Pull request merged**